### PR TITLE
Add Clearwater Controls Advanced Keypad and required components

### PR DESCRIPTION
**Pull request by FreddieChopin**
Local verification was successful (;

### CODE REVIEW

**Comment by FreddieChopin** [comment]
Please note that the official docs for DTDLv2 don't specify any limit on the number of fields for object. There is only a limit of 5 levels of depth, but this does not apply in this case.

https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/dtdlv2.md#object

Therefore I have absolutely no idea where does the limit of 30 fields come from.

**Comment by rido-min** [comment]
Hi @FreddieChopin ,

You are right, there is a missing comment on the DTDLv2 spec documenting the maximum number of fields. We have created this [issue](https://github.com/Azure/opendigitaltwins-dtdl/issues/91) to track. 

The workaround is to reduce the number of fields to 30. I'd like to better understand your motivation to have more than 30 fields, as we can re-evaluate this limit in the next revision of the language.

For clarification, the error comes from the DTDL Parser. If you would like to validate locally you can use the same tool we use in the PR validation pipeline, it's available [here](https://github.com/Azure/iot-plugandplay-models-tools/tree/dev/resolvers/dotnet).

Thanks for your submission
Rido

**Comment by FreddieChopin** [comment]
> I'd like to better understand your motivation to have more than 30 fields, as we can re-evaluate this limit in the next revision of the language.

I'd like to better understand your motivation to limit the number of fields to 30, as we may need to severely dumb-down our (seemingly very advanced) device to fit the ~ridiculous~ arbitrary constraints imposed by cutting-edge market-leading Azure IoT technology, which cannot accept a JSON object with 34 fields... (;

Sorry for that, but your question as well as the limit has really left me speechless for a brief moment (;

The device which we would like to add is used as a bridge between Azure and an industrial-grade controllers. This controller probably provides literally **hundreds** of values (I believe that number is pretty close to one thousand) which can be read and monitored. One device can be a bridge for several of such controllers, that's why the data is organized as an array of objects. For convenience and to limit the network traffic, we have grouped them into 7 "profiles", which can be individually turned on or off, also each has configurable period of refreshing. Some of these profiles are smaller, some of them are bigger, however I wouldn't call any of them "big". I really don't feel like artificially breaking profile nr 7 (the one with 34 fields) into several smaller ones, just because someone decided that "30 fields ought to be enough for anybody".

Again, sorry for that, but the only non-taunting answer to such question that I can provide is "because we need 34 fields in one object". Really, we just do. Maybe for IoIL (Internet of Intelligent Lightbulbs) 30 fields is like "a lot", but for us it's just not enough...

As this limit was never publicly documented and described anywhere, I propose to either remove it or make it more sane, say 999 fields. I really see no good reason why Azure couldn't handle more than 30.

**Comment by jrdouceur** [comment]
I can partly address your question as to why the limit is 30, although my memory is failing me on the specifics.  In general, the language limits are driven up by customer need and driven down by impact on services, so we try to select values that satisfy both.  When we chose limits for DTDL v2, we looked at models that had been uploaded to our public repo, looked at examples of devices we imagined might be converted to PnP, and talked with key customers.  Based on this data gathering, we selected values that seemed conservative given our level of knowledge at the time.  For Object fields in particular, we did not come across any cases of flat structures with dozens of fields; what seemed common was for developers to use hierarchy to organize their structures.  With 5 levels of complex schema hierarchy, DTDL allows 30^5 = 24.3 million fields for a single Property or Telemetry.  Of course, unless your hierarchy is extremely regular, you cannot take advantage of that full count, but you can go well beyond the hundreds or thousands of values that your controller has.

To break Rido's question into a few sub-questions:  Is there a reason these values are in a flat structure?  If so, how large do you anticipate that your flat structures are likely to be?  Hypothetically, if the language had a more complicated limit that allowed larger counts of fields per Object as long as the total hierarchical depth was lower, such that the total count of allowable values was kept to a few million, would this be helpful or just confusing?  A motivating example from a customer would be quite useful to us when determining any changes to limits in the next version of the DTDL language.

**Comment by FreddieChopin** [comment]
I've changed most of the problematic object into a map, as it also fits our requirements.

As for the discussion:
> With 5 levels of complex schema hierarchy, DTDL allows 30^5 = 24.3 million fields for a single Property or Telemetry.

This is exactly why I think the limit of 30 fields is just pointless. It's hard to say that the limit is justified by an attempt to minimize computation power required to process it, if you can actually have millions of entries by doing some simple workarounds. I'm sure that 24 milion entries in 5 levels of hierarchy are much harder to process than 34 elements in a single level, but I'm just an embedded guy working with microcontrollers (;

> Is there a reason these values are in a flat structure?

Any grouping there would be artificial, serving no purpose other than to work around the limit of 30 fields.

> Hypothetically, if the language had a more complicated limit that allowed larger counts of fields per Object as long as the total hierarchical depth was lower, such that the total count of allowable values was kept to a few million, would this be helpful or just confusing?

Why not a pretty high limit on total number of entries in the whole DTMI file? Say that the whole DTMI can have no more than 100k (or any value that is REALLY HIGH) "individual entries". By "individual entries" I mean both "leaf" and "compound" JSON entries, or you can count just the "leaf" entries or whatever. Just keep the limit really high. This would be more convenient for the customers (the limit would be just theoretical in almost all possible cases), yet this example 100k of entries is 250x smaller then the limit you currently have on a single object, so also convenient for your side. Someone wants a very deep hierarchy - fine. Someone wants a very big object with no hierarchy - also fine.

I think that reducing the depth and increasing the number of fields is just as artificial as the current situation - just the other way around. Especially because I think that the file I want to merge probably already uses 3 or 4 levels in some places... So I would risk an opinion, that this would create as many issues as it solved.

On the other hand, I really don't know why there has to be any (low) limit at all... There certainly are no such **LOW** limits for MQTT. Given that the MQTT topic can be 64kB long and the single MQTT message can be 256 MB long, the limits imposed by Azure look really pale... I wouldn't have any problem with a limits like the ones for MQTT, as I would consider them purely theoretical.

**Comment by jrdouceur** [comment]
I’m glad you found a satisfactory workaround.

We’ll be taking a comprehensive look at DTDL limits as part of the next version of the language.  Your feedback will be quite helpful.  Thanks much.

File: dtmi/uk/co/clearwatercontrols/deraggerplus/telemetry-1.json [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+  "@id": "dtmi:uk:co:clearwatercontrols:deraggerPlus:telemetry;1",
+  "@type": "Interface",
+  "displayName": "Deragger+ Telemetry",
+  "contents": [
+    {
+      "@type": "Telemetry",
+      "name": "telemetry1",
+      "displayName": "Deragger+ Telemetry Profile 1",
+      "schema": {
+        "@type": "Array",
```
Comment (rido-min) [inline dtmi/uk/co/clearwatercontrols/deraggerplus/telemetry-1.json:12]: Telemetry based Arrays are not yet supported by Azure IoT Central.

**Review by rido-min — APPROVED** [review]
Note : While DTDL supports types such as Arrays and Geospatial, IoT Solutions such as Azure IoT Central are in the process of onboarding these types. Your models might temporarily not work in Azure IoT Central.

**Comment by rido-min** [comment]
@FreddieChopin  Can you please update your PR branch (sync with origin/main), or allow maintainers to edit the PR so we can update it for you? Thank you.

**Comment by FreddieChopin** [comment]
@rido-min :
> Can you please update your PR branch (sync with origin/main)

I think I just did that (with the button here in this pull-request, so I hope this is the same as if I would do that manually).

**Comment by ridomin** [comment]
IoT Central will start supporting DTDL v2 before end of this year, DTDL v2 full support will be added in the first half of 2021.

**Comment by FreddieChopin** [comment]
@rido-min :
> IoT Solutions such as Azure IoT Central are in the process of onboarding these types. Your models might temporarily not work in Azure IoT Central.

Any info about when could be expect that to change? Is that something like "soon" or rather "some time in the future"?

**Comment by rido-min** [comment]
IoT Central will start supporting DTDL v2 before end of this year, DTDL v2 full support will be added in the first half of 2021.

**Comment by FreddieChopin** [comment]
@rido-min Two last question then - what can we expect to work before end of this year and for what we will have to wait for some more time? I hope the arrays are near the beginning of the queue, but are there any other features we used in this DTMI file that won't be supported this year (like map)? Does this limitation also affect the process of getting certification via https://certify.azure.com/products or maybe this will work just fine now?

**Comment by rido-min** [comment]
Before end of this year Central will support DTDL v2:

- Using DTMI identifiers
- No Component Scenarios
- Interfaces with Components
-- There is a known issue when referencing the same interface from 2 or more components.
- Basic semantic types support
- Import/Export with Central Application Templates

Azure IoT Central allows models with Maps, devices can send Maps data to IoTCentral and export the data out. Ability to model, visualize will be enabled in 2021/Q1

This does not impact your ability to get devices certified via https://certify.azure.com/products. Azure IoT Central will allow models with Arrays in 2001/Q1, there are no other features in your model that are impacted.

**Comment by FreddieChopin** [comment]
@rido-min thanks for the info!

**Pull request merged**